### PR TITLE
gh-121459: Deferred LOAD_ATTR (methods)

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Issue: gh-121459

### CODE REVIEW

File: Objects/object.c [hunk 1]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+                 tp->tp_name, name);
+
+    _PyObject_SetAttributeErrorContext(obj, name);
+    return 0;
+#else
+    PyObject *res = NULL;
+    int err = _PyObject_GetMethod(obj, name, &res);
+    *method = res == NULL ? PyStackRef_NULL : PyStackRef_FromPyObjectSteal(res);
+    return err;
+#endif
```
Comment (Fidget-Spinner) [inline Objects/object.c]: No clue why but if I remove this block, and the ifdef in general, it crashes on gil-build.

File: Objects/dictobject.c [hunk 2]
```diff
@@ -2324,6 +2324,27 @@ _PyDict_GetItemRef_KnownHash(PyDictObject *op, PyObject *key, Py_hash_t hash, Py
...
 
+int
+_PyDict_GetItem_KnownHash_StackRef(PyDictObject *op, PyObject *key, Py_hash_t hash, _PyStackRef *result)
+{
+#ifdef Py_GIL_DISABLED
+    Py_ssize_t ix = _Py_dict_lookup_threadsafe_stackref(op, key, hash, result);
+#else
+    PyObject *value;
+    Py_ssize_t ix = _Py_dict_lookup(op, key, hash, &value);
+    *result = PyStackRef_FromPyObjectSteal(value);
```
Comment (colesbury) [inline Objects/dictobject.c]: Doesn't `_Py_dict_lookup` return a borrowed reference? I don't see how the `PyStackRef_FromPyObjectSteal` is correct.

File: Include/internal/pycore_stackref.h [hunk 3]
```diff
@@ -149,8 +149,7 @@ PyStackRef_FromPyObjectNew(PyObject *obj)
 {
     // Make sure we don't take an already tagged value.
     assert(((uintptr_t)obj & Py_TAG_BITS) == 0);
-    assert(obj != NULL);
-    if (_Py_IsImmortal(obj) || _PyObject_HasDeferredRefcount(obj)) {
+    if (obj == NULL || _Py_IsImmortal(obj) || _PyObject_HasDeferredRefcount(obj)) {
```
Comment (colesbury) [inline Include/internal/pycore_stackref.h]: Let's avoid pushing the NULL check down here. It's going to hurt performance, and it doesn't match the default build behavior, which requires `obj != NULL`.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -5356,6 +5356,58 @@ find_name_in_mro(PyTypeObject *type, PyObject *name, int *error)
...
+    Py_ssize_t n = PyTuple_GET_SIZE(mro);
+    for (Py_ssize_t i = 0; i < n; i++) {
+        PyObject *base = PyTuple_GET_ITEM(mro, i);
+        PyObject *dict = lookup_tp_dict(_PyType_CAST(base));
+        assert(dict && PyDict_Check(dict));
+        if (_PyDict_GetItem_KnownHash_StackRef((PyDictObject *)dict, name, hash, result) < 0) {
+            *error = -1;
+            goto done;
+        }
+        if (!PyStackRef_IsNull(*result)) {
```
Comment (colesbury) [inline Objects/typeobject.c]: Use the return value from `_PyDict_GetItem_KnownHash_StackRef` instead of checking `*result`.

File: Python/bytecodes.c [hunk 5]
```diff
@@ -2000,17 +1998,17 @@ dummy_func(
...
-                    ERROR_IF(attr_o == NULL, error);
+                    ERROR_IF(PyStackRef_IsNull(*attr), error);
                     self_or_null = PyStackRef_NULL;
                 }
             }
             else {
                 /* Classic, pushes one value. */
-                attr_o = PyObject_GetAttr(PyStackRef_AsPyObjectBorrow(owner), name);
+                PyObject *attr_o = PyObject_GetAttr(PyStackRef_AsPyObjectBorrow(owner), name);
+                *attr = attr_o == NULL ? PyStackRef_NULL : PyStackRef_FromPyObjectSteal(attr_o);
```
Comment (colesbury) [inline Python/bytecodes.c]: Use an if-else statement instead of this code pattern. It's prone to MSVC compiler bugs.

https://github.com/python/cpython/pull/124085

File: Objects/typeobject.c [hunk 6]
```diff
@@ -5356,6 +5356,58 @@ find_name_in_mro(PyTypeObject *type, PyObject *name, int *error)
     return res;
 }
 
+
+static void
+find_name_in_mro_stackref(PyTypeObject *type, PyObject *name, int *error, _PyStackRef *result)
```
Comment (colesbury) [inline Objects/typeobject.c:5455]: `error` should be a return value, not an output argument

File: Objects/object.c [hunk 7]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+    PyErr_Format(PyExc_AttributeError,
+                 "'%.100s' object has no attribute '%U'",
+                 tp->tp_name, name);
+
+    _PyObject_SetAttributeErrorContext(obj, name);
+    return 0;
+#else
+    PyObject *res = NULL;
+    int err = _PyObject_GetMethod(obj, name, &res);
+    *method = res == NULL ? PyStackRef_NULL : PyStackRef_FromPyObjectSteal(res);
```
Comment (colesbury) [inline Objects/object.c]: Use an if-else statement instead of this code pattern (MSVC)

File: Objects/object.c [hunk 8]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
     return 0;
 }
 
+int
+_PyObject_GetMethodStackRef(PyObject *obj, PyObject *name, _PyStackRef *method)
```
Comment (colesbury) [inline Objects/object.c]: This should follow standard Python code practices where an error is indicated by a `-1` return (instead of `0`)

File: Objects/object.c (hunk 1 above)
Comment (colesbury) [reply Objects/object.c]: I think it's important that we figure that out.

File: Objects/object.c [hunk 9]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+
+    if (tp->tp_getattro != PyObject_GenericGetAttr || !PyUnicode_CheckExact(name)) {
+        *method = PyStackRef_FromPyObjectSteal(PyObject_GetAttr(obj, name));
+        return 0;
+    }
+
+    // Directly set it to that if a GC cycle happens, the descriptor doesn't get
+    // evaporated.
+    // This is why we no longer need a strong reference for this if it's
+    // deferred.
```
Comment (colesbury) [inline Objects/object.c]: I don't understand these two comments. Can you explain them?

File: Objects/object.c [hunk 10]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
     return 0;
 }
 
+int
+_PyObject_GetMethodStackRef(PyObject *obj, PyObject *name, _PyStackRef *method)
+{
```
Comment (colesbury) [inline Objects/object.c:1625]: It doesn't have to be in this PR, but we should replace all usages of `_PyObject_GetMethod` with `_PyObject_GetMethodStackRef`. Having two similar functions for the same purpose is going to be more difficult to maintain.

File: Objects/object.c [hunk 11]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+        *method = PyStackRef_FromPyObjectSteal(PyObject_GetAttr(obj, name));
+        return 0;
+    }
+
+    // Directly set it to that if a GC cycle happens, the descriptor doesn't get
+    // evaporated.
+    // This is why we no longer need a strong reference for this if it's
+    // deferred.
+    _PyType_LookupStackRef(tp, name, method);
+    _PyStackRef descr_st = *method;
```
Comment (colesbury) [inline Objects/object.c:1651]: It's confusing that `descr_st` and `*method` refer to the same data, and both are used below. I think it would be more clear if we avoid the aliasing and stick to `*method`.

File: Include/internal/pycore_dict.h [hunk 12]
```diff
@@ -118,7 +118,9 @@ PyAPI_FUNC(int) _PyDict_SetItem_KnownHash_LockHeld(PyDictObject *mp, PyObject *k
 PyAPI_FUNC(int) _PyDict_GetItemRef_KnownHash_LockHeld(PyDictObject *op, PyObject *key, Py_hash_t hash, PyObject **result);
 extern int _PyDict_GetItemRef_KnownHash(PyDictObject *op, PyObject *key, Py_hash_t hash, PyObject **result);
 extern int _PyDict_GetItemRef_Unicode_LockHeld(PyDictObject *op, PyObject *key, PyObject **result);
+extern int _PyDict_GetItem_KnownHash_StackRef(PyDictObject *op, PyObject *key, Py_hash_t hash, _PyStackRef *result);
```
Comment (colesbury) [inline Include/internal/pycore_dict.h]: `_PyDict_GetItemStackRef_KnownHash` to match the pattern used with other names

File: Objects/dictobject.c (hunk 2 above)
Comment (colesbury) [reply Objects/dictobject.c]: Also, can't we just use `_Py_dict_lookup_threadsafe_stackref` for both code paths? It's defined with what looks like essentially this implementation in the default build.

File: Objects/dictobject.c [hunk 13]
```diff
@@ -7054,6 +7093,79 @@ _PyObject_TryGetInstanceAttribute(PyObject *obj, PyObject *name, PyObject **attr
...
+    assert(keys != NULL);
+    Py_ssize_t ix = _PyDictKeys_StringLookup(keys, name);
+    if (ix == DKIX_EMPTY) {
+        *attr = PyStackRef_NULL;
+        return true;
+    }
+
+#ifdef Py_GIL_DISABLED
+    PyObject *value = _Py_atomic_load_ptr_acquire(&values->values[ix]);
+    *attr = PyStackRef_FromPyObjectNew(value);
```
Comment (colesbury) [inline Objects/dictobject.c]: `PyStackRef_FromPyObjectNew(value)` is not thread-safe here

File: Objects/dictobject.c [hunk 14]
```diff
@@ -7054,6 +7093,79 @@ _PyObject_TryGetInstanceAttribute(PyObject *obj, PyObject *name, PyObject **attr
...
+        // Caller needs to lookup from the dictionary
+        success = false;
+    }
+
+    Py_END_CRITICAL_SECTION();
+
+    return success;
+#else
+    PyObject *value = values->values[ix];
+    *attr = PyStackRef_FromPyObjectNew(value);
```
Comment (colesbury) [inline Objects/dictobject.c]: `value` might be `NULL`, right?

File: Objects/object.c [hunk 15]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+                PyStackRef_CLOSE(descr_st);
+                return 0;
+            }
+        }
+    }
+    PyObject *dict;
+    if ((tp->tp_flags & Py_TPFLAGS_INLINE_VALUES) &&
+        _PyObject_TryGetInstanceAttributeStackRef(obj, name, method)) {
+        if (!PyStackRef_IsNull(*method)) {
+            PyStackRef_XCLOSE(descr_st);
```
Comment (colesbury) [inline Objects/object.c:1675]: This overwrites `*method`, but not `descr_st`, so `descr_st` may store a stack reference that's not visible to the GC. That seems fragile and potentially a problem.

File: Objects/object.c [hunk 16]
```diff
@@ -1581,6 +1581,109 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
+        Py_DECREF(dict);
+    }
+
+    if (meth_found) {
+        *method = descr_st;
+        return 1;
+    }
+
+    if (f != NULL) {
+        *method = PyStackRef_FromPyObjectSteal(f(PyStackRef_AsPyObjectBorrow(descr_st), obj, (PyObject *)Py_TYPE(obj)));
```
Comment (colesbury) [inline Objects/object.c]: I'd like to avoid adding any more `PyStackRef_FromPyObjectSteal()` with potentially `NULL` arguments as the `NULL` check is adding overhead to every conversion.

File: Objects/object.c (hunk 9 above)
Comment (Fidget-Spinner) [reply Objects/object.c]: Woops it's outdated.

File: Objects/dictobject.c (hunk 2 above)
Comment (Fidget-Spinner) [reply Objects/dictobject.c]: You're right. Gonna use `_Py_dict_lookup_threadsafe_stackref` for both.

File: Objects/dictobject.c (hunk 13 above)
Comment (Fidget-Spinner) [reply Objects/dictobject.c]: We need the TryXGetRef pattern right?

File: Objects/object.c (hunk 8 above)
Comment (Fidget-Spinner) [reply Objects/object.c]: This behavior is mirroring `_PyObject_GetMethod`

File: Objects/object.c (hunk 15 above)
Comment (Fidget-Spinner) [reply Objects/object.c:1675]: We want to restore *method to `descr_st` after this in some cases (see below). I guess the safe thing would be to convert it to a strong ref?

File: Objects/typeobject.c (hunk 6 above)
Comment (Fidget-Spinner) [reply Objects/typeobject.c:5455]: This mirros find_name_in_mro, but I guess since this is completely private it's fine to fix it.

File: Objects/object.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Objects/object.c]: I think I figured out why. Should be something related to https://github.com/python/cpython/pull/124301 (ie more off by ones somewhere).

**Comment by Fidget-Spinner** [comment]
@colesbury gentle ping to respond to the review comments please

File: Objects/object.c (hunk 8 above)
Comment (colesbury) [reply Objects/object.c]: Oh, right. I forgot about that.

File: Objects/dictobject.c (hunk 13 above)
Comment (colesbury) [reply Objects/dictobject.c]: Yes, but with `_PyStackRef`. I think something like:

```c
static inline _PyStackRef
_Py_TryXGetStackRef(PyObject **ptr)
{
    PyObject *value = _Py_atomic_load_ptr(ptr);
    if (value == NULL) {
        return PyStackRef_NULL;
    }
    if (_Py_IsImmortal(value) || _PyObject_HasDeferredRefcount(value)) {
        return (_PyStackRef){ .bits = (uintptr_t)value | Py_TAG_DEFERRED };
    }
    if (_Py_TryIncrefCompare(ptr, value)) {
        return PyStackRef_FromPyObjectSteal(value);
    }
    return PyStackRef_NULL;
}
```

File: Objects/object.c (hunk 15 above)
Comment (colesbury) [reply Objects/object.c:1675]: I don't think you need both `*method` and `descr_st` if you refactor these function.

Alternatively take in a second `_PyStackRef *` to store the temporary values in a place that is visible to the GC.

Comment (Fidget-Spinner) [reply Objects/object.c:1675]: I think the second approach needs to wait for @mpage 's PR on adding an extra stack slot?

Comment (Fidget-Spinner) [reply Objects/object.c:1675]: Oh we don't, we can use `self_or_null`.

Comment (Fidget-Spinner) [reply Objects/object.c:1675]: Ok nevermind, this does indeed need Matt's fix. Because we unconditionally write to self_or_null now.

File: Include/internal/pycore_object.h [hunk 17]
```diff
@@ -623,6 +623,24 @@ _Py_TryXGetRef(PyObject **ptr)
     return NULL;
 }
 
+// This belongs here rather than pycore_stackref.h because including pycore_object.h
+// there causes a circular include.
+static inline _PyStackRef
+_Py_TryXGetStackRef(PyObject **ptr)
```
Comment (markshannon) [inline Include/internal/pycore_object.h:629]: Please add clear comments as to what this does.
What is it "trying" to do? What happens if it fails? etc.

File: Include/internal/pycore_object.h [hunk 18]
```diff
@@ -816,7 +834,8 @@ extern int _PyObject_StoreInstanceAttribute(PyObject *obj,
                                             PyObject *name, PyObject *value);
 extern bool _PyObject_TryGetInstanceAttribute(PyObject *obj, PyObject *name,
                                               PyObject **attr);
-
+extern bool _PyObject_TryGetInstanceAttributeStackRef(PyObject *obj, PyObject *name,
```
Comment (markshannon) [inline Include/internal/pycore_object.h:837]: Likewise. Anything with "try" in the name needs to be carefully documented.

File: Include/internal/pycore_object.h [hunk 19]
```diff
@@ -863,6 +882,7 @@ PyAPI_FUNC(PyObject*) _PyObject_LookupSpecialMethod(PyObject *self, PyObject *at
 extern int _PyObject_IsAbstract(PyObject *);
 
 PyAPI_FUNC(int) _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method);
+PyAPI_FUNC(int) _PyObject_GetMethodStackRef(PyObject *obj, PyObject *name, _PyStackRef *method, _PyStackRef *spare);
```
Comment (markshannon) [inline Include/internal/pycore_object.h:885]: Why does this have a "spare" parameter?

File: Include/internal/pycore_object.h [hunk 20]
```diff
@@ -900,6 +920,8 @@ PyAPI_DATA(int) _Py_SwappedOp[];
 
 extern void _Py_GetConstant_Init(void);
 
+extern void _PyType_LookupStackRef(PyTypeObject *type, PyObject *name, _PyStackRef *result);
```
Comment (markshannon) [inline Include/internal/pycore_object.h:923]: Can't this fail? It should probably return an `int` indicating success/failure/error.

File: Python/bytecodes.c [hunk 21]
```diff
@@ -2054,20 +2054,19 @@ dummy_func(
             #endif  /* ENABLE_SPECIALIZATION */
         }
 
-        op(_LOAD_ATTR, (owner -- attr, self_or_null if (oparg & 1))) {
+        op(_LOAD_ATTR, (owner -- attr[1], self_or_null[1] if (oparg & 1))) {
```
Comment (markshannon) [inline Python/bytecodes.c:2057]:
```suggestion
        op(_LOAD_ATTR, (owner -- attr, self_or_null if (oparg & 1))) {
```

You don't need the arrays. The code generator should spill the values to the in-memory stack if necessary.

File: Python/compile.c [hunk 22]
```diff
@@ -1335,6 +1335,11 @@ optimize_and_assemble_code_unit(struct compiler_unit *u, PyObject *const_cache,
         goto error;
     }
 
+    /* Reserve an extra word on the stack to ensure there is space for uops to
```
Comment (markshannon) [inline Python/compile.c:1338]: Why is this necessary?

**Review by markshannon — COMMENTED** [review]
This seems to add a *lot* of complexity.
`LOAD_ATTR` is already the slow path, do you have performance numbers?
Without a significant speedup, this doesn't seem worth it. 

For future reference, for large PRs like this, could you make sure there is specific issue where the change can be discussed and justified, rather adding them to a more general (and vague) issue like https://github.com/python/cpython/issues/121459

**Comment by Fidget-Spinner** [comment]
@markshannon I'm tempted to wait for Matt to land his LOAD_ATTR specialization PR then take a look again. It may be possible that specialization makes this PR less useful.

**Comment by Fidget-Spinner** [comment]
I'm closing this because https://github.com/python/cpython/pull/128164 addresses practically everything.
